docs: clean up some confusion about state management and pages. (#10553)
sveltejs/kit: documentation/docs/20-core-concepts/50-state-management.md

If you're not using SSR (and can guarantee that you won't need to use SSR in future) then you can safely keep state in a shared module, without using the context API.

> ## Component and Page states are preserved

Reviewer: ```suggestion
## Component and page state is preserved
```

If you're not using SSR (and can guarantee that you won't need to use SSR in future) then you can safely keep state in a shared module, without using the context API.

## Component and Page states are preserved

> > In SvelteKit, pages and layouts are considered components and follow the same rules and limitations as normal Svelte components.

Reviewer 2: ```suggestion
> In SvelteKit, pages and layouts are components and follow the same rules and limitations as normal Svelte components.
```

> When you navigate around your application, SvelteKit reuses existing layout and page components. For example, if you have a route like this...

Reviewer 2: I just realised these two blocks are basically saying the same thing, we should consolidate them

Author: Moved some wording around making it clear that layout, page and any components on the page are being reused (this also fits with the code a bit lower where we use `#key` to force a specific component to rerender).

Also included a quick note about `onMount` and `onDestroy`, nor sure we need a code sample for those though.

<div>{@html data.content}</div>
```

> ...then navigating from `/blog/my-short-post` to `/blog/my-long-post` won't cause the layout, page and any other components within to be destroyed and recreated. Instead the `data` prop (and by extension `data.title` and `data.content`) will reactively change (as it would with a regular Svelte component) and, because the code isn't rerunning, lifecycle methods like `onMount` and `onDestroy` won't rerun and `estimatedReadingTime` won't be recalculated.

Reviewer: ```suggestion
...then navigating from `/blog/my-short-post` to `/blog/my-long-post` won't cause the layout, page and any other components within to be destroyed and recreated. Instead the `data` prop (and by extension `data.title` and `data.content`) will update (as it would with any other Svelte component) and, because the code isn't rerunning, lifecycle methods like `onMount` and `onDestroy` won't rerun and `estimatedReadingTime` won't be recalculated.
```

> ...then navigating from `/blog/my-short-post` to `/blog/my-long-post` won't cause the components (layout, page and any components on the page) to be destroyed and recreated. Instead the `data` prop (and by extension `data.title` and `data.content`) will reactively change (as it would with a regular Svelte component) and, because the code isn't rerunning, `estimatedReadingTime` won't be recalculated nor will lifecycle methods like `onMount` and `onDestroy` run.

Reviewer 3: This isn't 100% correct, as we don't know how downstream components interact with each other - maybe an if block makes a different component appear? I'm not sure how to best phrase it to remove that ambiguity.

Reviewer 2: ```diff
- won't be recalculated nor will lifecycle methods like `onMount` and `onDestroy` run.
+ won't be recalculated and lifecycle methods like `onMount` and `onDestroy` won't rerun.

Author: I removed some of the text and elaborated more in the part where the example with a `#key` block is to emphasize that components that have already been rendered are not re-rendered.

<div>{@html data.content}</div>
```

> ...then navigating from `/blog/my-short-post` to `/blog/my-long-post` won't cause the components (layout and page) to be destroyed and recreated. Instead the `data` prop (and by extension `data.title` and `data.content`) will reactively change (as it would with a regular Svelte component) and, because the code isn't rerunning, `estimatedReadingTime` won't be recalculated and lifecycle methods like `onMount` and `onDestroy` won't rerun.

Reviewer 4: ```suggestion
...then navigating from `/blog/my-short-post` to `/blog/my-long-post` won't cause the layout, page and any other components within to be destroyed and recreated. Instead the `data` prop (and by extension `data.title` and `data.content`) will reactively change (as it would with a regular Svelte component) and, because the code isn't rerunning, `estimatedReadingTime` won't be recalculated and lifecycle methods like `onMount` and `onDestroy` won't rerun.
```

> If your code in `onMount` and `onDestroy` has to run again after navigation you can use [aferNavigate](modules#$app-navigation-afternavigate) and [beforeNaviate](modules#$app-navigation-beforenavigate) respectively.

> Instead, we need to make the value [_reactive_](https://learn.svelte.dev/tutorial/reactive-assignments):

Reviewer 4: This sentence is referring to the `estimatedReadingTime` variable, the flow is broken by the change.

<div>{@html data.content}</div>
```

> ...then navigating from `/blog/my-short-post` to `/blog/my-long-post` won't cause the component to be destroyed and recreated. Instead the `data` prop (and by extension `data.title` and `data.content`) will change as happens with regular Svelte component, and because the code isn't rerunning, `estimatedReadingTime` won't be recalculated.

Reviewer 2: ```suggestion
...then navigating from `/blog/my-short-post` to `/blog/my-long-post` won't cause the component to be destroyed and recreated. Instead the `data` prop (and by extension `data.title` and `data.content`) will reactively change (as it would with a regular Svelte component) and, because the code isn't rerunning, `estimatedReadingTime` won't be recalculated.
```

</script>
```

> > If your code in `onMount` and `onDestroy` has to run again after navigation you can use [aferNavigate](modules#$app-navigation-afternavigate) and [beforeNaviate](modules#$app-navigation-beforenavigate) respectively.

Reviewer: ```suggestion
> If your code in `onMount` and `onDestroy` has to run again after navigation you can use [afterNavigate](modules#$app-navigation-afternavigate) and [beforeNavigate](modules#$app-navigation-beforenavigate) respectively.
```

</script>
```

> Reusing components like this means that things like sidebar scroll state are preserved, and you can easily animate between changing values. However, because only the state changes all other components on the page will also behave the same and will not remount if they are already visible on the page, only their props will change. In the case that you do need to completely destroy and remount a component on navigation, you can use this pattern:

Reviewer 4: This seems a bit verbose.

```suggestion
Reusing components like this means that things like sidebar scroll state are preserved, and you can easily animate between changing values. In the case that you do need to completely destroy and remount a component on navigation, you can use this pattern:
```

How about keeping it as is and addressing child components in the paragraph above?

{/key}
```

> > Instead of `onMount` and `onDestroy` you can use [aferNavigate](modules#$app-navigation-afternavigate) and [beforeNaviate](modules#$app-navigation-beforenavigate) respectively.

Reviewer 3: I think it would be better to have this up there after the "no will lifecycle methods run" sentence.

Author: Yeah, I wasn't sure where to put it to not break the 'flow' of going into the reactivity example, should it then be before the code sample or after ?

Author: Moved it
